File: go/internal/storage/cold_storage.go
package storage

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"ducker/internal/model"
)

type ColdStorage struct {
	RootDir string
}

func NewColdStorage(rootDir string) *ColdStorage {
	return &ColdStorage{RootDir: rootDir}
}

func (c *ColdStorage) GetManifest(tenantID string) (model.Manifest, error) {
	p := filepath.Join(c.RootDir, tenantID, "manifest.json")
	b, err := os.ReadFile(p)
	if err != nil {
		return model.Manifest{}, err
	}
	var m model.Manifest
	if err := json.Unmarshal(b, &m); err != nil {
		return model.Manifest{}, err
	}
	return m, nil
}

func (c *ColdStorage) GetBloom(tenantID, segment string) (map[string]any, error) {
	p := filepath.Join(c.RootDir, tenantID, "_bloom", segment+".bloom.json")
	b, err := os.ReadFile(p)
	if err != nil {
		return nil, err
	}
	var d map[string]any
	if err := json.Unmarshal(b, &d); err != nil {
		return nil, err
	}
	return d, nil
}

func (c *ColdStorage) ListTenants() ([]string, error) {
	entries, err := os.ReadDir(c.RootDir)
	if err != nil {
		return nil, err
	}
	out := make([]string, 0)
	for _, e := range entries {
		if e.IsDir() && strings.HasPrefix(e.Name(), "tenant-") {
			out = append(out, e.Name())
		}
	}
	return out, nil
}

File: node/src/storage/coldStorage.js
import { readFile, readdir } from 'node:fs/promises';
import path from 'node:path';

export class ColdStorage {
  constructor(rootDir) {
    this.rootDir = rootDir;
  }

  resolve(relativePath) {
    return path.join(this.rootDir, relativePath);
  }

  async getManifest(tenantId) {
    const p = path.join(this.rootDir, tenantId, 'manifest.json');
    return JSON.parse(await readFile(p, 'utf-8'));
  }

  async getBloom(tenantId, segmentName) {
    const p = path.join(this.rootDir, tenantId, '_bloom', `${segmentName}.bloom.json`);
    return JSON.parse(await readFile(p, 'utf-8'));
  }

  async listTenants() {
    const entries = await readdir(this.rootDir, { withFileTypes: true });
    return entries.filter((e) => e.isDirectory() && e.name.startsWith('tenant-')).map((e) => e.name);
  }
}
